build_tools/linux_python_dist_split.py

```diff
@@ -0,0 +1,524 @@
+
+The issue arises because a standard Unix library is symlinked like this:
+
+```
+libamd_comgr.so -> libamd_comgr.so.2
+libamd_comgr.so.2 -> libamd_comgr.so.2.9
+libamd_comgr.so.2.9
+```
+
+But the SONAME is `libamd_comgr.so.2` -- a symlink that we would like to
+avoid (and the only file that matters at runtime). This makes some sense in an
+OS context where you may have multiple patch versions of a library co-existing
+and being arbitrated by symlink. But it doesn't fit our constraints. So we
+"rotate" all link farms: In the normal packages, we only materialize the SONAME
+variant (the middle one above) as a real file. In devel packages with shared
+libraries that were materialized in a runtime package, we symlink back to the
+runtime package version.
+
+The ROCM and LLVM distribution also includes a number of executable symlink,
+many of the form:
+
+```
+clang*
+amdclang -> clang
+```
+
+Critically, all such symlinks are relative. We replace them with a stub
+executable that we compile as part of the script which can perform the
+dynamic path computation and exec() the intended target.
+
+There are other alterations we have to make to accomodate our uniquely split
+layout on disk.
+
+Fat and Thin Target Packages
+----------------------------
+Target specific packages (aka `libraries`) can either be fat or thin:
+
+* Fat target packages will contain independent host+device code in their own
+  dedicated tree joined together. For multi-target packaging scenarios, there
+  will be one per family, and the top-level `rocm-sdk` package will contain
+  the meta-data mapping actual devices to subsets. Such packages will be
+  named with a `_target_family` in the name. For development or single-host
+  use, we support distribution of a single tree with no target-specific
+  indirection (i.e. does not contain the `_target_family`).
+* Thin target packages contain target-neutral host code in a non target specific
+  directory tree, and then all device code/kernel databases are located
+  separately and loaded via an interposer.
+
+Thin target packages are currently in the conceptual phase but are intended to
+be the final default packaging mechanism for multi-target builds.
+
+Development Package
+-------------------
+The `-devel` package is unique because it is the catch-all for any distributed
+files that are materialized in one of the above packages. For anything that is
+materialized in a prior package, a relative symlink is generated across the
+root directory to the named package directory. Since the development package
+contains symlink farms, it cannot be included verbatim in a Python wheel (which
+would try to materialize the symlinks). Instead, we *embed* it as a .tar.xz
+file in a built wheel at Python package construction time. Then code that
+needs to get the ROCM `cmake_prefix_path()` or run any tools will dynamically
+decompress it to the HOME directory, setup root links, and run from there.
```

Comment:
> to the HOME directory

Is that configurable? I wonder if developers would be surprised by that behavior

The following "Note that most users never will need the development package" does help with that concern though. I guess "development" here is similar to CMake's `-Wno-dev` which is for _the author of the CMakeLists.txt files_. That is, if you are working on distributable builds _of TheRock_, you may care about this. If you just want the latest outputs of the package builds or want to make changes to a subproject, they are not for you.

Reply:
Well, it doesn't exist yet, so "sure" :)